### ISSUE TITLE
Full-text analyzer config prevents Lucene 5+ update

### ISSUE THREAD

**Issue by adamretter** [eXist-db/exist#1160]
We have updated Lucene to the latest version of the 4.x branch, we cannot yet more to 5+ becuase of a mismatch between how eXist allows ananlyzers to be specified and how Lucene now uses analyzers for Indexing.

eXist currently allows you to specify different analyzers, per-field per-indexed Document. Lucene 5+ only allows either a single analyzer to be specified per Index, or per-field (on all documents within the same Index).

The relevant Lucene issue is - https://issues.apache.org/jira/browse/LUCENE-6212

We have 4 or more different options as I see it:
1. Have multiple Lucene indexes, one Index per-configured collection. This allows the analyzers used per field to be fixed for the fields in all of the documents of a specific Collection. We can do this using [PerFieldAnalyzerWrapper](https://lucene.apache.org/core/4_2_0/analyzers-common/org/apache/lucene/analysis/miscellaneous/PerFieldAnalyzerWrapper.html). At search time we would have to query multiple Lucene indexes, we could use Lucene's https://lucene.apache.org/core/5_0_0/core/org/apache/lucene/index/MultiReader.html to achieve this. This could actually have some advantages, for example, being able to Full-Text index multiple collections in parallel and also being able to multi-thread the full text lookup (one thread per-collection). However we would need to understand the overhead of amalgamating the results during a query. This approach would not require any changes to eXist's collection.xconf.

2. We stay with a single index, but we use [PerFieldAnalyzerWrapper](https://lucene.apache.org/core/4_2_0/analyzers-common/org/apache/lucene/analysis/miscellaneous/PerFieldAnalyzerWrapper.html) in combination with naming each Field based on the analyzer. The issue here is that if we have two indexes on the field `@thing` from different collections which use different analyzers, then we need to use anazlyzer specific fields per-document, e.g. Lucene document fields called `@thing:my-analyzer1` or `@thing:my-analyzer2`. At Query time we would need to query the correct field for the analyzer configured on the collection field, so we would have to look it up based on the collection configs, or pass the complexity onto the end user.

3. We completely modify the way that full-text indexes are created in eXist so that they are based around the Lucene analyzer rather than the Collection. So a user would define a specific named full-text index (e.g. `my-index-1`) that uses a certain analyzer, at query time they would specify which index it is that they wish to query, e.g. `ft:query("my-index-1", $some-node, <query>...</query>)`

4. We do (1) but we have a single Lucene Index per-analyzer rather than per-collection, based on the assumption that there will be less Analyzers than Collections; this would give us just a few large Lucene indexes rather than potentially many small Lucene indexes, which should hopefully make the performance acceptable.

There are probably combinations of the above that are possible too. I quite like the idea of at least trying (4) and seeing how that works out.

**Comment by dizzzz** [comment]
From my side.... I do not really have a preferred solution.

From a 'user' perspective it would be a bad thing if -after an upgrade- my index-configurations would not work anymore. It would be acceptable though if we could provide scripts to upgrade xconf files

**Comment by joewiz** [comment]
Thanks for this analysis @adamretter! I think only your number 3 should be discounted, since as @dizzzz suggests, this would be not only a configuration change but a change to all code. Two questions:

Do any of the solutions offer a better opportunity to align with the [XQuery and XPath Full Text Facility recommendation](https://www.w3.org/TR/xpath-full-text-30/)? Not that this should be a dealbreaker, but it would be great to be able to offer even limited support of this very nice syntax, and the only standard to speak of in the XQuery world.

Are there any implications for the new range index in this move to Lucene 5+, since the new range index uses Lucene too?

**Comment by adamretter** [comment]
@joewiz None of the options have an impact on `XQuery and XPath Full Text Facility recommendation` as far as I can see.

Regards the Range Index, there are no immediate implications.

**Comment by joewiz** [comment]
@adamretter Thanks, great to hear on both fronts.

**Comment by alebastrov** [comment]
How about to upgrade to 6.x ? Looks likeIndexWriter allows to use several anilizers (with method addDocuments). Just asking because 6.x and 5.x branches of lucene has many speed optimizations, which would be nice to have.
P.S. we use search query like
~~~
collection('/db/MyApp/data/')/DOCUMENT[ft:query(., $fquery)]
~~~
and it may take about 10 seconds to perform, I believe that newest lucene version could perform much faster

**Comment by dizzzz** [comment]
@alebastrov yes, probably, *but* I understood that Lucene is changing the public API quite often, making it difficult to follow it; I believe that even some important features are now missing..... (on the 5.x API)

**Comment by shabanovd** [comment]
@dizzzz lucene always have good [migration guide](https://lucene.apache.org/core/6_0_0/MIGRATE.html)

**Comment by adamretter** [comment]
@dizzzz @shabanovd having migrated eXist from 4.x through several 5.x versions, my personal opinion is that the Lucene Migration guides are severely lacking and the Lucene documentation in general is so thin as to be almost useless for any  serious project.

@alebastrov I have started on a 6.x migration but haven't found the time to complete it

**Comment by shabanovd** [comment]
Well, I had no issue during migration. And I found new lucene fix old problem with simpler resolution.

**Comment by joewiz** [comment]
@adamretter For your 5.x work, did you decide on one of the four approaches in your original post? What do you think the outlook is for this move? Do you need more feedback here?

**Comment by joewiz** [comment]
Lucene 5.5 fixes a bug affecting our version of Lucene, which is that when performing a fuzzy search, exact matches are scored lower than fuzzy matches. See the question posed at https://stackoverflow.com/questions/46197136/xquery-fuzzy-search-in-exist-db/46198421#46198421 and my reply with links to where this was fixed in Lucene 5.5.

Is an upgrade to Lucene 5.5+ or 6.x still in the works?

**Comment by adamretter** [comment]
@joewiz I am happy to complete this work. However, first I need to get the id/idref fixed that @wolfgangmm has proposed merged in. After that I need to understand how the new Collation stuff that @wolfgangmm implemented should work, as that isn't directly compatible with Lucene 5+.

Personally, my priorities are to get the locking and recovery corruption issues solved first. However, if there was a commercial customer with a pressing need... ;-)

**Comment by joewiz** [comment]
@adamretter Thanks, good to know. Could you point me to the id & idref issue/discussion? And if there's an issue or PR for the collation stuff, could you share a link to that too?

**Comment by adamretter** [comment]
@joewiz I am referring to the stuff raised by Wolfgang in the last community call (11/09/2017).

**Comment by joewiz** [comment]
This issue was referenced in https://stackoverflow.com/questions/46197136/xquery-fuzzy-search-in-exist-db/46198421.

**Comment by duncdrum** [comment]
closed with #6146
